### ISSUE TITLE
CNN으로 텍스트 분류하기

### ISSUE THREAD

**Issue by nureeee** [nureeee/DeepLearning#10]
```
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Embedding, Dropout, Conv1D, GlobalMaxPooling1D # Input Channel 만큼의 필터의 결과물에서 제일 큰거
from tensorflow.keras.layers import Dense, Input, Flatten, Concatenate
```

* 임베딩 차원, dropout 비율, 필터 개수 정의
```
embedding_dim = 128
dropout_prob = 0.5
num_filters = 128 # 커널마다의 필터가 128개
```
* 레이어 구성

```
inputs = Input(shape=(max_len, ))

net = Embedding(vocab_size, embedding_dim, input_length=max_len)(inputs)
net = Dropout(dropout_prob)(net)

net.shape
```

```
# 합성곱 연산 수행
conv_blocks = [] # 합성곱의 결과를 모아낼 리스트 ( 합성곱 - 풀링 - 평탄화까지 진행한 결과를 저장)

filters = [3, 4, 5]

for filter_size in filters:
  # 합성곱 연산
  conv = Conv1D(
      filters=num_filters, # 필터의 개수
      kernel_size = filter_size, # 커널의 크기
      padding='valid',
      activation='relu',
      strides=1)(net) # 알쏭달쏭 파이썬세상

  # 풀링
  conv = GlobalMaxPooling1D()(conv)

  # 풀링된 결과를 평탄화 시켜서 배열에 넣기
  conv = Flatten()(conv)
  conv_blocks.append(conv)

# conv_blocks에 2개 이상의 결과물이 들어있으면 그대로 추가, 그게 아니면(1개만 있는 경우) 첫 번째 블록만 추가.
net = Concatenate()(conv_blocks) if len(conv_blocks) > 1 else conv_blocks[0]

# Fully Connected
net = Dense(128, activation='relu')(net)

# 출력층 설계
model_output = Dense(1, activation='sigmoid')(net)

model = Model(inputs, model_output)
model.compile(
    loss='binary_crossentropy',
    optimizer='adam',
    metrics=['acc']
)

model.summary()
```

```
es = EarlyStopping(
    monitor='val_loss',
    mode='min',
    verbose=1,
    patience=4
)

mc = ModelCheckpoint(
    "CNN_Text_model.h5", # 케라스에서 제공하는 레이어만으로 만들었으면 .h5로 저장 - load_model
                         # 커스텀된 레이어를 사용했으면 ckpt로 저장 - model.load_weights로 불러옴
    monitor="val_acc",
    mode='max',
    verbose=1,
    save_best_only=True)
```

```
model.fit(
    X_train_padded,
    y_train,
    batch_size = 64,
    epochs=10,
    validation_data=(X_test_padded, y_test),
    callbacks=[es, mc])
```
